File: 2term/jvmWorld/2lab_kotlin_btree/src/main/kotlin/CustomTypeTrait.kt
interface CustomTypeTrait {

    val typeName: String

    val typeExampleRepresentation: String

    fun createRandomInstance(): CustomTypeTrait

    override fun toString(): String

    fun serializeToString(): String

    fun deserializeFromString(ss: String): CustomTypeTrait?

    fun compare(other: CustomTypeTrait): Int
}

File: 2term/jvmWorld/1lab_scala_binaryTree/src/main/scala/CustomTypeTrait.scala
trait CustomTypeTrait {
  def typeName: String

  def typeExampleRepresentation: String

  def createRandomInstance(): this.type

  def toString: String

  def serializeToString: String

  def deserializeFromString(ss: String): Option[this.type]

  def compare(other: CustomTypeTrait): Int
}
